File: briar-android/src/main/java/org/briarproject/briar/android/contact/InboxThreadMerger.java
package org.briarproject.briar.android.contact;

import org.briarproject.nullsafety.NotNullByDefault;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

@NotNullByDefault
final class InboxThreadMerger {

	private static final Comparator<InboxThreadItem> BY_LATEST_ACTIVITY =
			Comparator.comparingLong(InboxThreadItem::getLatestActivityMillis)
					.reversed()
					.thenComparing(InboxThreadItem::getStableId);

	static List<InboxThreadItem> merge(List<ContactListItem> briarItems,
			List<TelegramInboxThreadItem> telegramItems) {
		List<InboxThreadItem> items =
				new ArrayList<>(briarItems.size() + telegramItems.size());
		for (ContactListItem item : briarItems) {
			items.add(new BriarInboxThreadItem(item));
		}
		items.addAll(telegramItems);
		items.sort(BY_LATEST_ACTIVITY);
		return items;
	}

	static void sort(List<InboxThreadItem> items) {
		items.sort(BY_LATEST_ACTIVITY);
	}
}

File: briar-android/src/main/kotlin/org/briarproject/briar/android/contact/InboxThreadMerger.kt
package org.briarproject.briar.android.contact

object InboxThreadMerger {

	private val byLatestActivity = compareByDescending<InboxThreadItem> {
		it.latestActivityMillis
	}.thenBy {
		it.stableId
	}

	@JvmStatic
	fun merge(
		briarItems: List<ContactListItem>,
		connectorItems: List<ConnectorInboxThreadItem>,
	): List<InboxThreadItem> {
		val items = ArrayList<InboxThreadItem>(
			briarItems.size + connectorItems.size
		)
		briarItems.mapTo(items) { BriarInboxThreadItem(it) }
		items.addAll(connectorItems)
		items.sortWith(byLatestActivity)
		return items
	}

	@JvmStatic
	fun sort(items: MutableList<InboxThreadItem>) {
		items.sortWith(byLatestActivity)
	}
}
